ovn-tester/cms/openstack/openstack.py

```diff
@@ -0,0 +1,724 @@
+        specification.
+
+        :param project: Project to which external network will be added
+        :param external_network: External network that will be connected.
+        :return: Logical Router Port that connects to the external network.
+        """
+
+        ls_port, lr_port = self._add_router_port_external_gw(
+            external_network.neutron_net, project.router
+        )
+        self.external_port = lr_port
+        gw_net = DualStackSubnet(netaddr.IPNetwork("0.0.0.0/0"))
+
+        # XXX: ovsdbapp does not allow setting external IDs to static route
+        # XXX: Setting 'policy' to "" throws "constraint violation" error in
+        #      logs because ovsdbapp does not allow not specifying policy.
+        #      However, the route itself is created successfully with no
+        #      policy, the same way Neutron does it.
+        self.nbctl.route_add(project.router, gw_net, lr_port.ip, "")
+
+        gw_nodes = self._get_gateway_chassis(external_network.num_gw_nodes)
+        for index, chassis in enumerate(gw_nodes):
+            self.nbctl.lr_port_set_gw_chassis(
+                lr_port, chassis.container, index + 1
+            )
+
+        project.ext_net = external_network
+
+        return lr_port
+
+    def add_internal_network_to_project(
+        self, project: Project, snat_port: Optional[LRPort] = None
+    ) -> None:
+        """Provision internal net to the Project for guest communication.
+
+        If 'snat_port' is provided. A SNAT rule will be configured for traffic
+        exiting this network.
+
+        :param project: Project to which internal network will be added.
+        :param snat_port: Gateway port that should SNAT outgoing traffic.
+        :return: None
+        """
+        int_net_name = f"int_net_{project.uuid}"
+        int_net = self._create_project_net(int_net_name, 1442)
+
+        int_net_port = self._add_metadata_port(int_net, project.uuid)
+        security_group = self._create_default_security_group()
+        neutron_int_network = NeutronNetwork(
+            network=int_net,
+            ports={int_net_port.uuid: int_net_port},
+            name=int_net_name,
+            security_group=security_group,
+        )
+
+        self._add_network_subnet(network=neutron_int_network)
+        self._assign_port_ips(neutron_int_network)
+        project.int_net = neutron_int_network
+
+        self._add_router_port_internal(
+            neutron_int_network, project.router, project
+        )
+
+        snated_network = DualStackSubnet(int_net.cidr)
+        if snat_port is not None:
+            self.nbctl.nat_add(project.router, snat_port.ip, snated_network)
+
+    def add_vm_to_project(self, project: Project, vm_name: str):
+        compute = self.select_worker_for_port()
+        vm_port = self._add_vm_port(
+            project.int_net, project.uuid, compute, vm_name
+        )
+        compute.bind_port(vm_port)
```

Comment:
we may want to also set mtu_request, as os-vif ovs port implementation does. This may affect behavior of OVN in some cases. (but this can be left to later.)

https://github.com/openstack/os-vif/blob/0a0dec37e42e220ffa750650027714bd7289faa1/vif_plug_ovs/ovsdb/ovsdb_lib.py#L68